### ISSUE TITLE
[GACODE follow-up] Add TGLF transport and NN surrogate backends with verified model resolution

### ISSUE THREAD

**Issue by HongSik-Yun-Fusion** [VEST-Tokamak/vaft#553]
## Summary

Follow up on #550 by extending the shared GACODE profile/runtime infrastructure to support **TGLF-based turbulent transport** and **TGLF-NN-style neural-network surrogates**.

The goal is not to add an isolated TGLF executable wrapper or a standalone ML utility. Instead, VAFT should define a reusable local transport contract that can be consumed by both native TGLF and accelerated surrogate backends while preserving model provenance, validity information, and IMAS/OMAS semantics.

Conceptually:

```text
canonical scientific state
    equilibrium + core_profiles
              |
              v
     GACODEProfile / input.gacode
              |
              v
        local TGLF state
          /         \
         v           v
   native TGLF    NN surrogate
                     |
                TGLF-NN / future
          \          /
           v        v
        TransportResult
              |
              v
       validation / comparison
              |
              v
        IMAS core_transport
```

The NN path should be treated as an **accelerated implementation of the same transport-model contract**, not as a separate canonical ML data model.

Reference implementations / model sources:

- #550
- GACODE: https://github.com/gafusion/gacode
- TurbulentTransport.jl: https://github.com/ProjectTorreyPines/TurbulentTransport.jl
- FUSE.jl: https://github.com/ProjectTorreyPines/FUSE.jl

---

## Motivation

#550 intentionally establishes the GACODE suite boundary first:

```text
shared GACODE profile/runtime contract
        |
        +-- NEO
        +-- TGLF       future
        +-- CGYRO      future
```

This issue should build directly on that infrastructure rather than introducing a second kinetic-profile representation or a disconnected `vaft.code.tglf` implementation.

TurbulentTransport.jl already demonstrates a useful architecture in which first-principles transport models and fast NN surrogates consume the same TGLF inputs and return comparable flux outputs. FUSE then uses these through transport actors rather than treating the NN as an unrelated ML subsystem.

VAFT should adopt the same core idea while remaining Python-native and independent of the Julia/FUSE runtime:

```text
IMAS / OMAS scientific state
        |
        v
reproducible local transport state
        |
        +-- external native solver
        |
        `-- portable surrogate inference
        |
        v
solver/model-native result
        |
        v
validation + standardized projection
```

---

# 1. Extend the #550 GACODE ownership boundary

Do not create unrelated top-level packages such as:

```text
vaft.code.tglf
vaft.ml.tglfnn
```

Prefer extending:

```text
vaft/code/gacode/
├── _runtime.py
├── _profiles.py
├── ...
│
├── neo/
│
├── tglf/
│   ├── __init__.py
│   ├── _input.py
│   ├── _runtime.py
│   ├── _output.py
│   ├── _types.py
│   │
│   └── surrogate/
│       ├── _registry.py
│       ├── _resolver.py
│       ├── _onnx.py
│       └── _types.py
│
└── cgyro/              # future
```

Exact file names may change, but preserve the conceptual ownership:

```text
GACODEProfile
    = shared GACODE interoperability state

TGLFInput
    = local TGLF physics input

TGLFResult / TransportResult
    = model-native transport result

surrogate
    = TGLF-compatible accelerated inference backend
```

---

# 2. Define a typed `TGLFInput` local transport contract first

Do not begin with ONNX inference.

First define a typed local state that can be consumed by both native TGLF and surrogate models.

Conceptually:

```python
@dataclass
class TGLFInput:
    rho: float

    # local equilibrium / geometry
    q: float
    shear: float
    ...

    # kinetic state
    ne: float
    te: float
    ...

    # gradients
    a_lne: float
    a_lte: float
    ...

    # species
    ions: ...

    # rotation / ExB
    ...
```

The actual field inventory and normalization must be derived from the real TGLF input contract.

Input construction should follow:

```text
equilibrium + core_profiles
        |
        v
GACODEProfile
        |
        v
TGLFInput(rho)
```

Reuse existing VAFT local-equilibrium / Miller-geometry infrastructure where appropriate.

At minimum, explicitly handle:

- radial coordinate;
- local geometry;
- q and magnetic shear;
- electron density and temperature;
- ion species composition;
- density and temperature gradients;
- Zeff / charge information;
- toroidal rotation;
- ExB shear;
- electromagnetic quantities;
- TGLF normalization;
- source equilibrium/core-profile time;
- interpolation provenance.

Do not fabricate missing required inputs merely to make a model runnable.

Preserve the distinction between:

```text
measured / reconstructed input
derived input
caller-supplied modelling assumption
missing required input
```

---

# 3. Add a native TGLF backend

Reuse the GACODE runtime resolution established by #550.

Conceptual API:

```python
result = vaft.code.gacode.tglf.run(
    tglf_input,
    backend="native",
)
```

or an equivalent API consistent with the final `vaft.code` conventions.

The native backend should own:

- TGLF executable resolution;
- deterministic working-directory lifecycle;
- `input.tglf` generation;
- solver execution;
- failure propagation;
- native output parsing;
- normalization recovery;
- executable/version provenance;
- input identity;
- run configuration identity.

---

# 4. Add TGLF-NN as a surrogate of the same physics contract

Do not create a separate scientific input representation for ML inference.

Maintain:

```text
TGLFInput
   |
   +------------ native TGLF
   |
   `------------ TGLF-NN
```

Conceptual API:

```python
result = vaft.code.gacode.tglf.run(
    tglf_input,
    backend="surrogate",
    model="sat3_em_mastu+nstx_azf-1_withnegD",
)
```

or, if implementation separation is clearer:

```python
result = vaft.code.gacode.tglf.surrogate.run(
    tglf_input,
    model=...,
)
```

The exact public API can be decided during implementation, but native and surrogate outputs should be comparable through the same transport-result semantics whenever physically defensible.

This does not conflict with #550's decision to keep Sauter and NEO as distinct computational identities. Sauter and NEO are different neoclassical models, while TGLF-NN is explicitly trained as a surrogate for a TGLF-family transport contract.

---

# 5. Do not vendor NN weights in the VAFT repository

Do not commit large pretrained model files to VAFT, including via Git LFS.

When public upstream artifacts are available, VAFT should store only a small **model registry / manifest**.

Conceptually:

```yaml
sat3_em_mastu_nstx:
  backend: tglfnn
  format: onnx

  upstream:
    repository: ProjectTorreyPines/TurbulentTransport.jl
    ref: <pinned-ref>
    path: models/...

  artifact:
    sha256: ...
    size: ...

  physics:
    sat_rule: 3
    electromagnetic: true
    devices:
      - MAST-U
      - NSTX

  training_domain:
    ...
```

Preserve, where available:

- canonical model identifier;
- upstream repository;
- pinned commit/tag/ref;
- artifact path / URL;
- SHA-256;
- expected byte size;
- model format;
- required external-data files;
- training device/domain;
- saturation rule;
- electrostatic/electromagnetic identity;
- output channels;
- model/version provenance;
- license / redistribution information.

The actual model weights remain external artifacts.

---

# 6. Deterministic model resolution: local/environment first, remote fallback

Model lookup should follow a deterministic precedence such as:

```text
1. explicit caller path / model object
        |
2. configured local model directory
        |
3. TurbulentTransport repository from environment
        |
4. validated local VAFT model cache, if enabled
        |
5. upstream public artifact download
```

If a local TurbulentTransport checkout is configured through an environment variable, reuse it before attempting network access.

Conceptually:

```bash
export TURBULENTTRANSPORT_ROOT=/path/to/TurbulentTransport.jl
```

The final variable name should be checked against existing VAFT external-code environment conventions before being fixed.

When the root exists, search the corresponding upstream model directories first.

If the checkout contains only a Git LFS pointer, use its declared:

```text
oid sha256
size
```

as integrity metadata for resolving the real artifact.

Where compatible with upstream semantics, also consider supporting:

```text
TURBULENTTRANSPORT_MODELS_REF
```

or an equivalent VAFT-side override for reproducible model-ref selection.

---

# 7. Support memory-first loading for public ONNX models

Users should not need to clone the whole upstream repository or manually install NN weights merely to run a public ONNX model.

For a self-contained ONNX artifact, prefer:

```text
model registry
     |
     v
HTTPS download
     |
     v
bytes in memory
     |
     v
SHA-256 verification
     |
     v
onnxruntime.InferenceSession(bytes)
     |
     v
process-level session cache
```

In other words, the normal remote path should be able to use:

```text
remote artifact
    -> verified bytes
    -> memory
    -> ONNX Runtime
```

without requiring a permanent local copy.

This keeps the VAFT repository and user environment free of unnecessary large model artifacts.

---

# 8. Allow verified materialization for ONNX external-data models

Memory-only execution must not become an absolute requirement.

Some ONNX models may require external sidecars such as:

```text
model.onnx
model.onnx.data
```

which ONNX Runtime resolves relative to the model file.

Therefore use:

```text
single-file ONNX
    -> verified in-memory loading

ONNX + external data
    -> verified temporary/cache directory
       model.onnx
       model.onnx.data
       ...
```

Every downloaded artifact must be hash-validated.

The architectural rule is:

> do not vendor model weights into VAFT

not:

> model bytes must never touch the filesystem

---

# 9. Optional persistent cache

Repeatedly downloading large models on every process launch is also undesirable.

Design the resolver so both modes can be supported:

```text
memory
    remote -> RAM/session only

cache
    remote -> verified user cache -> inference
```

Conceptually:

```python
load_model(
    "...",
    cache="memory",
)
```

or:

```python
load_model(
    "...",
    cache_dir=...,
)
```

The final API can differ.

Persistent cache identity must be tied to at least:

```text
model id
upstream ref
sha256
```

Do not treat filename alone as a sufficient cache identity.

---

# 10. Keep network access outside normal VAFT import/runtime

These must remain offline-safe:

```python
import vaft
import vaft.code.gacode
```

Remote access should occur only when the user explicitly requests a model that cannot be resolved locally.

Offline failure should clearly report:

```text
model not found locally
+
remote retrieval unavailable
+
expected model id / searched local paths
+
how to provide an explicit path or configured repository root
```

Do not automatically clone repositories or install packages.

---

# 11. Keep ONNX Runtime optional

Do not make a heavy ML runtime an unconditional VAFT dependency.

The intended dependency layering is approximately:

```text
vaft
vaft + GACODE external executables
vaft + ONNX surrogate runtime
```

VAFT and the native GACODE adapter must remain importable without `onnxruntime` installed.

Only surrogate execution should require the optional runtime.

---

# 12. Treat model validity / extrapolation as first-class output

A neural network can return a numerical value outside its training domain, so a plain flux array is insufficient.

Where available, preserve information such as:

```python
prediction.validity
prediction.extrapolation
prediction.uncertainty
prediction.model
prediction.provenance
```

At minimum record:

- model identifier;
- model SHA-256;
- upstream repository/ref;
- training devices;
- training bounds;
- input variables outside bounds;
- ensemble uncertainty, when available;
- surrogate fidelity;
- reference solver identity.

Do not treat an in-domain prediction and a strong extrapolation as equivalent results in later database or plotting workflows.

This is especially important when applying public models to VEST or other low-aspect-ratio devices outside their original training distribution.

---

# 13. Preserve `TransportResult` before IMAS projection

Do not write parser/inference output directly into `core_transport` and discard model-native meaning.

Follow the same general principle as `NeoResult` in #550:

```text
TGLFResult / TransportResult
    = model-native scientific source of truth

IMAS projection
    = standardized/queryable representation
```

The result should preserve, where applicable:

```text
rho
species identity

particle flux
energy / heat flux
momentum flux

normalization
TGLF saturation rule
electromagnetic configuration

backend:
    native
    surrogate

surrogate model identity
training-domain status
uncertainty
provenance
```

Unsupported outputs must remain absent / unevaluated rather than being zero-filled.

---

# 14. Map defensible results into IMAS `core_transport`

After the native result contract is stable, audit IMAS/OMAS mappings for quantities such as:

```text
core_transport
    electron particle transport / flux
    electron energy transport / flux
    ion particle transport / flux
    ion energy transport / flux
    momentum-related quantities
```

Mapping must be based on physical definition, not similar field names.

For every mapped quantity, verify:

- gyro-Bohm vs physical normalization;
- radial flux sign convention;
- heat vs energy flux definition;
- convective contribution;
- species indexing;
- radial coordinate;
- flux-surface averaging;
- momentum normalization.

---

# 15. Native TGLF <-> TGLF-NN validation

A core deliverable is direct comparison from the same local input:

```text
native TGLF
      |
      +----> flux_native

TGLF-NN
      |
      +----> flux_surrogate
```

Recommended validation hierarchy:

```text
upstream TGLF-NN reference/example
        |
        v
VAFT ONNX inference parity
        |
        v
same TGLFInput
native TGLF <-> TGLF-NN
        |
        v
multi-radius profile comparison
        |
        v
device-specific scientific assessment
```

Candidate comparison quantities include:

- electron energy flux;
- ion energy flux;
- electron particle flux;
- ion particle flux;
- momentum flux;
- radial RMS error;
- relative error;
- uncertainty;
- training-domain violations.

---

# 16. Initial device/reference strategy

Do not assume a VEST-specific NN model exists.

Start with public models and reference cases whose provenance and training domain are well defined.

Relevant public model families include device-specific and multi-machine models for systems such as:

```text
DIII-D
MAST-U
NSTX
MAST-U + NSTX
multi-machine combinations
```

For spherical-tokamak validation, prioritize MAST-U/NSTX-relevant public models first.

VEST application should proceed only after:

```text
input domain audit
        |
        v
in-domain / extrapolation assessment
        |
        v
native TGLF comparison
```

A successful numerical prediction alone must not be interpreted as model qualification for VEST.

---

# 17. Relationship to FUSE / TurbulentTransport

Use FUSE/TurbulentTransport as reference architecture and pretrained-model source, but do not require:

```text
Julia installation
FUSE installation
TurbulentTransport.jl runtime
PythonCall / PyJulia bridge
```

The VAFT inference path should remain Python-native.

Preserve the important common contract:

```text
same physical TGLF inputs
        |
native physics model / surrogate
        |
same interpretable flux outputs
```

This also makes future VAFT-vs-FUSE comparison possible.

---

# 18. Tests

## Unit tests

The default CI path must not require network access.

Cover at least:

- `TGLFInput` construction;
- unit/normalization conversion;
- missing-required-input behavior;
- model registry parsing;
- explicit local-path resolution;
- environment-root resolution;
- SHA-256 verification;
- checksum mismatch rejection;
- mocked remote download;
- in-memory ONNX loading;
- external-data materialization;
- offline error behavior;
- training-bound detection;
- result/provenance serialization.

Do not commit large production model weights as normal unit-test fixtures.

Use a small synthetic ONNX fixture if needed.

## Optional integration tests

When network/external-code dependencies are available, test:

```text
public upstream ONNX
        |
        v
download + verification
        |
        v
VAFT inference
```

and:

```text
native TGLF
    <- same input ->
TGLF-NN
```

---

# Acceptance criteria

## Architecture

- [ ] Reuse the #550 `GACODEProfile / input.gacode` infrastructure.
- [ ] Do not introduce an unrelated top-level TGLF package.
- [ ] Provide a typed `TGLFInput` or equivalent local transport contract.
- [ ] Native TGLF and surrogate backends share the same scientific input semantics.
- [ ] Do not make the surrogate representation a new canonical kinetic-state model.

## Native TGLF

- [ ] Native TGLF input/run/output adapter works.
- [ ] Executable/version/input provenance is preserved.
- [ ] VAFT remains importable without GACODE installed.

## Surrogate

- [ ] At least one public TGLF-NN ONNX model can be executed from Python.
- [ ] Julia/FUSE runtime is not required.
- [ ] NN runtime is an optional dependency.
- [ ] Model identity and SHA-256 are preserved in prediction provenance.
- [ ] Training-domain violations are detected and exposed.

## Model artifacts

- [ ] NN weights are not vendored in the VAFT repository.
- [ ] Explicit local paths are supported.
- [ ] Environment-variable-based TurbulentTransport checkout resolution is supported.
- [ ] Public upstream artifacts can be fetched when not available locally.
- [ ] Every remote artifact is cryptographically verified.
- [ ] Self-contained ONNX supports memory-first inference.
- [ ] ONNX external data supports verified temporary/cache materialization.
- [ ] Persistent cache validates model/ref/hash identity.
- [ ] Import and unrelated VAFT workflows never require network access.

## Validation

- [ ] VAFT ONNX inference is checked against an upstream reference inference.
- [ ] Native TGLF and TGLF-NN can be compared using the same `TGLFInput`.
- [ ] Flux comparison and extrapolation/uncertainty information can be represented together.
- [ ] At least one ST-relevant public model/reference case is evaluated.

## Standardization

- [ ] Preserve the model-native transport result before IMAS projection.
- [ ] Review defensible `core_transport` mappings.
- [ ] Document normalization, sign, species, and radial-coordinate semantics.

---

# Out of scope

This issue does **not** include:

- training new neural networks;
- a VAFT-native ML training pipeline;
- copying public NN weights into the VAFT repository;
- reimplementing FUSE;
- requiring the Julia/FUSE runtime;
- CGYRO integration;
- full GKNN integration;
- QLNN;
- FINN;
- ModeID-NN;
- TGYRO flux matching;
- full transport time evolution;
- automatic profile evolution;
- campaign-scale production execution;
- a universal ML backend API for all future surrogates;
- automatically accepting out-of-training-domain predictions as qualified physical results.

These can follow after the TGLF/TGLF-NN contract and validation path are stable.

---

# Intended endpoint

When complete, the GACODE branch of VAFT should conceptually reach:

```text
                    canonical VAFT state
                equilibrium + core_profiles
                           |
                           v
                     GACODEProfile
                     /           \
                    /             \
                 NEO            TGLFInput
                                  |
                         +--------+--------+
                         |                 |
                         v                 v
                    native TGLF        TGLF-NN
                         |                 |
                         +--------+--------+
                                  |
                                  v
                         TransportResult
                                  |
                    +-------------+-------------+
                    |                           |
                    v                           v
                validation                core_transport
```

This should extend the GACODE interoperability layer established in #550 from neoclassical modelling into turbulent transport and accelerated surrogate modelling without changing VAFT's canonical scientific-state boundaries.
